Clarify support of UTF-8 and other encodings (#232)
graphql/graphql-over-http: spec/GraphQLOverHTTP.md

For details of the shapes of these JSON payloads, please see
[Request](#sec-Request) and [Response](#sec-Response).

If the media type in a `Content-Type` or `Accept` header does not

> include encoding information, then `utf-8` MUST be assumed (e.g.
> `Content-Type: application/graphql-response+json; charset=utf-8`).

Reviewer: This example doesn't seem to fit any more
```suggestion
include encoding information, then `utf-8` MUST be assumed (e.g.
for header `Content-Type: application/graphql-response+json`, UTF-8 encoding would be assumed).
```

Author: :+1:

If the client supplies an `Accept` header, the client SHOULD include the media
type `application/graphql-response+json` in the `Accept` header.

> Note: From 1st Jan 2025, every _server_ and _client_ MUST support

Author: ```suggestion
Note: From 1st Jan 2025, every _server_ and _client_ must support
```

`application/graphql-response+json` media type (as indicated by the `Accept`
header).

> A server MUST support the `utf-8` encoding for requests which accept a JSON media type,
> such as the `application/json` or `application/graphql-response+json` media types.

Author: FYI, I tried to word the paragraph in the same manner as the previous two paragraphs, which use the phrase "accept", as this does relate to the `Accept` HTTP header.

If I understand your concern, there's "Accept" which is the HTTP header regarding the response by the server, versus "accept" which is the server accepting a request from the client.  Is that correct?  If so, should we change any of the prior paragraphs as well?  Perhaps I should simply add the phrase "(as indicated by the `Accept` header)"?

Here's some ideas for this paragraph:

```suggestion
A server MUST support the `utf-8` encoding for requests which accept a JSON media type,
such as the `application/json` or `application/graphql-response+json` media types (as indicated by
the `Accept` header).
```

```suggestion
A server MUST support the `utf-8` encoding for requests which return a supported JSON media type,
such as the `application/json` or `application/graphql-response+json` media types (as indicated by
the `Accept` header).
```

I'm open to suggestions.

> such as the `application/json` or `application/graphql-response+json` media types.

Reviewer: Thinking about it more, I don't think we need nor want this paragraph. If you support `application/graphql-response+json` right now but only when using `ISO-8859-1` then that's okay. After the watershed you must support it with UTF-8 encoding (I believe this is stated elsewhere?) but that's a specific media-type and encoding combo we explicitly dictate. If you support `application/flibble+json; encoding=ISO-8859-1` I don't think we should dictate this must also be UTF-8.

> such as the `application/json` or `application/graphql-response+json` media types.

Author: Sounds good; take a look now.

> such as the `application/json` or `application/graphql-response+json` media types.

Author: So I think this puts us at:

POST requests' `Content-Type`:
- `application/json` must be supported
- `application/json` must be assumed when not specified
- `utf-8` must be assumed for `application/json` media type without encoding specified
- No explicit statement mandating support of `application/json` with `utf-8` encoding, but this is implied by the following statement:

> Request encoding with media type application/json is supported by every compliant server.

`Accept` header:
- `application/json` must be supported indefinitely
- `application/json` must be assumed until waterfall
- `application/graphql-response+json` must be supported after waterfall
- `application/graphql-response+json` must be assumed after waterfall
- `utf-8` is assumed for requests that accept `application/json` or `application/graphql-response+json` without encoding specified
- `application/graphql-response+json; charset=utf-8` must be supported after waterfall
